### ISSUE TITLE
Virtualmin LDAP:  Failed to save user : Failed to modify user in LDAP database : mailAlternateAddress: attribute type undefined

### ISSUE THREAD

**Issue by abclution** [virtualmin/virtualmin-gpl#230]
I setup Virtualmin + LDAP according to the documentation.

In my virtualmin config I have additionally selected

- Add mail attribute to LDAP users?  Yes using mail and mailAlternateAddress

![image](https://user-images.githubusercontent.com/10130237/99291406-6f993500-2848-11eb-9d63-70d43f6237e4.png)

When trying to save a user that has additional email addresses (aliases) I get this error:

Failed to save user : Failed to modify user in LDAP database : mailAlternateAddress: attribute type undefined

Now, the description for the Virtualmin setting indicates
"This is useful if you are using the same LDAP database with a mail server which uses that attribute to figure out where to deliver email."

I'm using Postfix which I am guessing, probably does not use mailAlternateAddress

My question is, where exactly is creating those attributes for the schema the responsibility of?
Its not part of the LDAP configuration instructions of Virtualmin, if I am reading correct.

I selected the option in the Virtualmin config as I expected it to be a cleaner way of delineating the primary and secondary addresses for the users in LDAP. But the schema doesn't seem to be correct for this option.

**Comment by seven0f** [comment]
OS: Debian 12; Latest version of Virtualmin

With the following option selected, new users cannot be added to domains if an additional email address is supplied.

Virtualmin > Virtualmin Configuration > Advanced options > Add mail attribute to LDAP users > Yes using mail and mailAlternateAddress

When adding a new user with additional email address(es), this error is returned:

Failed to save user : Failed to add user to LDAP database : mailAlternateAddress: attribute type undefined

When adding a new user without additional email address(es), the user is created and the LDAP object for the user contains a populated mail attribute.

After changing the configuration above to ..> Add mail attribute to LDAP users > Yes using only mail attribute

When adding a new user with additional email address(es), the user is created and the LDAP object contains a mail attribute for the 'primary' email address and for each additional email address.

Additionally, an object is created under dc=Virtual,dc=domain,dc=tld for each email address as expected.

As far as I could find, mailAlternateAddress is an attribute of the MailGroup object class, but I cannot find a schema that actually includes that class. It isn't defined in any of the schema/ldif under /etc/ldap/schema.

I found references to MailGroup here: https://ldapwiki.com/wiki/Wiki.jsp?page=MailGroup; and reference to mailAlternateAddress here: https://docs.oracle.com/cd/E28280_01/reference.1111/e10035/schema_attrs.htm#OIMUR0801. The latter links mailAlternateAddress to RFC 2256, but RFC 2256 (X.500) doesn’t mention mailAlternareAddress and neither do the RFCs that supercede RFC 2256.

**Comment by seven0f** [comment]
Perhaps some text can be added to the help tooltip noting that the attribute must be availble in your site's LDAP schema(s).

**Comment by jcameron** [comment]
The assumption for these attributes is that you would have added them to your LDAP schema.

However, in a typical Virtualmin setup, mailAlternateAddress isn't used to deliver email - instead we use Postfix LDAP integration, which has it's own schema.

**Comment by neosonic2** [comment]
Sorry to resurface an old issue, but I'm curious as to how one might go about adding the **mailAlternateAddress** attribute to their LDAP schema?

I changed the **Add mail attribute to LDAP users** setting to use both **mail** and **mailAlternateAddress** attributes, but I get the same error as the original commenter when I try to save an email user account on one of my virtual servers. I'm running an OpenLDAP server but I don't see a definition for the attribute anywhere in the built-in schemas and don't know how I would add it.

For reference, my OpenLDAP server is using the dynamic configuration format (i.e. the existing schemas that make up the server are all in the LDIF format).

**Comment by jcameron** [comment]
Are you certain that the `mailAlternateAddress` attribute is in the schema your LDAP server is using?

**Comment by neosonic2** [comment]
I don't think it is - but that's actually what I'm wondering about, how would I go about adding the attribute to my schema? I searched through the schema files in _/etc/ldap/schemas_ and there was nothing found. I would assume those are the base schemas for the LDAP server I'm running.

I've done some Google searching but am unable to find a concrete answer to this question. The best I can find is to switch from OpenLDAP to a server that already has this attribute in one of its built-in schemas.

**Comment by chris001** [comment]
https://forum.virtualmin.com/t/add-mail-attribute-to-ldap-users-not-adding-e-mail-to-mail-attribute/34581/4

**Comment by neosonic2** [comment]
Thanks for the link you posted, but unfortunately that doesn’t resolve my issue. None of the schemas in /etc/ldap/schemas contain definitions for the mailAlternateAddress attribute, and the link you posted does not provide information about how to define this attribute in the LDAP server.

**Comment by cowsonindustries** [comment]
You're right, the attribute does not appear in any of the schemas. I was unable to track down with certainty its origin and I eventually got around the problem by giving up trying to find a schema that included it, or create my own to add in to LDAP. I hope this matches your expectations.

https://forum.virtualmin.com/t/users-with-additional-email-addresses-not-create-when-add-mail-attribute-to-ldap-users-yes-using-mail-and-mailalternateaddress-is-selected-in-advanced-virtualmin-options/122590

**Comment by neosonic2** [comment]
Thanks for the info. Unless one of the Virtualmin devs can share more light on this, I may end up giving up on this effort as well and just use the mail attribute only in Virtualmin (rather than both mail and mailAlternateAddress), but it really would be nice to use both and Virtualmin supports it which means the devs must have known there would be some way to get the attribute added to one’s LDAP server.

The closest I could find to a schema that includes it was this schema file <https://github.com/debops/debops/blob/master/ansible/roles/slapd/files/etc/ldap/schema/debops/mailservice.schema>, but it requires another schema to work properly and seems to be tailored more for an Ansible-based project I’m not using.

**Comment by cowsonindustries** [comment]
With the information provided in that schema, I think you should be able to create your own schema extension and add it to your LDAP configuration. I don't know in detail how you would do that but suspect you could view the files under /etc/ldap/schemas and work out how to format your file. See https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/5/html/deployment_guide/s1-ldap-files-schemas#s1-ldap-files-schemas for example. Best wishes.

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
